### PR TITLE
Use path.Join to concatenate prefix and folder string (#2643)

### LINKED ISSUE #2643: Clarification on http-prefix ending with a trailing slash
Hello,

Today I upgraded from Cortex 0.7 where I used `http.prefix=/`. While testing things, I understood that the behavior of `http.prometheus-http-prefix` (and `http.alertmanager-http-prefix`) is different from what `http.prefix` used to do.

Using the value `/` or `/something/` will lead to 404 errors that are hard to spot.

I understand that Cortex is doing strict matching so it might not make sense to normalize URLs to avoid match `http://domain.com//api` as `http://domain.com/api`, however I do think it might be a good idea to warn the user about this behavior. Either in the documentation or as a startup warning `<...-http-prefix> has a trailing slash, this might not behave as you expect`.
When using CLI, this is even more confusing as the expected value to match on the root is nothing, which looks like this: `--http.prometheus-http-prefix=`. This is prone to being deleted as it doesn't looks like we are setting anything.

I have not found anything regarding this decision (I might have missed it) so I am asking here.

Is this the expected behavior? Does it deserve a warning?

Best regards,

### PR DESCRIPTION
**Pull request by ChinYing-Li**
Signed-off-by: ChinYing-Li <chinying.li@mail.utoronto.ca>

**What this PR does**:
Properly concatenates url

**Which issue(s) this PR fixes**:
Fixes #2643.
I am not sure whether this counts as a BUGFIX or a ENHANCEMENT.
Any suggestion is appreciated!

**Checklist**
- [ ] Tests updated
- [ ] Documentation added
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: pkg/api/api.go [hunk 1]
```diff
@@ -65,8 +66,8 @@ func (cfg *Config) RegisterFlags(f *flag.FlagSet) {
 
 // RegisterFlagsWithPrefix adds the flags required to config this to the given FlagSet with the set prefix.
 func (cfg *Config) RegisterFlagsWithPrefix(prefix string, f *flag.FlagSet) {
-	f.StringVar(&cfg.AlertmanagerHTTPPrefix, prefix+"http.alertmanager-http-prefix", "/alertmanager", "HTTP URL path under which the Alertmanager ui and api will be served.")
-	f.StringVar(&cfg.PrometheusHTTPPrefix, prefix+"http.prometheus-http-prefix", "/prometheus", "HTTP URL path under which the Prometheus api will be served.")
+	f.StringVar(&cfg.AlertmanagerHTTPPrefix, path.Join(prefix, "http.alertmanager-http-prefix"), "/alertmanager", "HTTP URL path under which the Alertmanager ui and api will be served.")
+	f.StringVar(&cfg.PrometheusHTTPPrefix, path.Join(prefix, "http.prometheus-http-prefix"), "/prometheus", "HTTP URL path under which the Prometheus api will be served.")
```
Comment (jtlisi) [inline pkg/api/api.go:70]: **nit:** These are flags, no URL routes. The prefix is meant to be concatenated as it currently is.

**Review by jtlisi — CHANGES_REQUESTED** [review]
One nit, otherwise this seems reasonable to me.

**Comment by pstibrany** [comment]
Should we fix `NewQuerierHandler` in handlers.go too? (at the end, where `cfg.PrometheusHTTPPrefix` and `cfg.LegacyHTTPPrefix` are used)

**Comment by pracucci** [comment]
> Should we fix `NewQuerierHandler` in handlers.go too? (at the end, where `cfg.PrometheusHTTPPrefix` and `cfg.LegacyHTTPPrefix` are used)

I think so. @ChinYing-Li do you mind taking a look, please?

File: pkg/api/api.go [hunk 2]
```diff
@@ -302,16 +303,16 @@ func (a *API) RegisterRulerAPI(r *ruler.API) {
...
 	// Legacy Ruler API Routes
-	a.RegisterRoute(a.cfg.LegacyHTTPPrefix+"/rules", http.HandlerFunc(r.ListRules), true, "GET")
-	a.RegisterRoute(a.cfg.LegacyHTTPPrefix+"/rules/{namespace}", http.HandlerFunc(r.ListRules), true, "GET")
-	a.RegisterRoute(a.cfg.LegacyHTTPPrefix+"/rules/{namespace}/{groupName}", http.HandlerFunc(r.GetRuleGroup), true, "GET")
-	a.RegisterRoute(a.cfg.LegacyHTTPPrefix+"/rules/{namespace}", http.HandlerFunc(r.CreateRuleGroup), true, "POST")
-	a.RegisterRoute(a.cfg.LegacyHTTPPrefix+"/rules/{namespace}/{groupName}", http.HandlerFunc(r.DeleteRuleGroup), true, "DELETE")
-	a.RegisterRoute(a.cfg.LegacyHTTPPrefix+"/rules/{namespace}", http.HandlerFunc(r.DeleteNamespace), true, "DELETE")
+	a.RegisterRoute(path.Join(a.cfg.LegacyHTTPPrefix, "/rules"), http.HandlerFunc(r.ListRules), true, "GET")
+	a.RegisterRoute(path.Join(a.cfg.LegacyHTTPPrefix, "/rules/{namespace}"), http.HandlerFunc(r.ListRules), true, "GET")
+	a.RegisterRoute(path.Join(a.cfg.LegacyHTTPPrefix, "/rules/{namespace}/{groupName})"), http.HandlerFunc(r.GetRuleGroup), true, "GET")
```
Comment (pracucci) [inline pkg/api/api.go:312]: A `)` slipped in.

```suggestion
	a.RegisterRoute(path.Join(a.cfg.LegacyHTTPPrefix, "/rules/{namespace}/{groupName}"), http.HandlerFunc(r.GetRuleGroup), true, "GET")
```

**Review by pracucci — COMMENTED** [review]
Good job! I left a comment about a typo which slipped in and then LGTM.

**Review by pracucci — APPROVED** [review]
LGTM, thanks!

**Review by jtlisi — APPROVED** [review]
LGTM

**Comment by pstibrany** [comment]
Thanks!

**Pull request merged**